scripts/generate_states.py

```diff
@@ -0,0 +1,108 @@
+
+# Copyright The Mbed TLS Contributors
+# SPDX-License-Identifier: Apache-2.0
+#
+# Licensed under the Apache License, Version 2.0 (the "License"); you may
+# not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+import sys
+from mbedtls_dev import build_tree
+
+OUTPUT_TEMPLATE = '''\
+/* Automatically generated by generate_states.py. DO NOT EDIT. */
+
+#include "common.h"
+
+#if defined(MBEDTLS_DEBUG_C)
+
+#include "mbedtls/ssl.h"
+#include "ssl_misc.h"
+
+const char *mbedtls_ssl_str_state( mbedtls_ssl_states state )
+{{
+    const char * state_to_str[]=
+    {{
+        {state_list}
+    }};
+
+    if( state > ( sizeof( state_to_str )/sizeof( state_to_str[0]) - 1 ) ||
+        state_to_str[ state ] == NULL )
+    {{
+        return "UNKOWN_STATE";
+    }}
+    return state_to_str[ state ];
+}}
+#endif /* MBEDTLS_DEBUG_C */
+/* End of automatically generated file. */
+
+'''
+def get_enum_definition(name, header='include/mbedtls/ssl.h'):
+    '''
+    Find enum definition in header file
+    '''
+    with open(header, 'r') as f:
+        print_flag = False
+        _temp = ''
+        for line in f:
+            line = line.rstrip()
+            if 'typedef enum' in line:
+                print_flag = True
+                _temp = ''
+            if not print_flag:
+                continue
+            _temp += line + '\n'
+            if line.endswith(';'):
+                if name in _temp:
+                    return _temp
+                _temp = ''
+                print_flag = False
+    return ''
+def generate_str_states_table():
+    '''
+    Generate enum translation table
+    '''
+    enum_definition = get_enum_definition('mbedtls_ssl_states')
+    start = enum_definition.find('{')+1
+    end = enum_definition.find('}')
+    enum_definition = enum_definition[start:end]
+    for line in enum_definition.splitlines():
+        fields = line.strip().split(',')
+        state = fields[0]
+        if line.startswith('#'):
+            yield line
+            continue
+        if not state:
+            continue
+        yield '{indent}[{state}] = "{state}",'.format(state=state, indent=' '*8)
+def generate_states(target_dir):
+    '''
+        Generate function for translating mbedtls_ssl_state to string.
+    '''
+
+    with open(os.path.join(target_dir, 'state.c'), 'w') as f:
+        f.write(OUTPUT_TEMPLATE.format(
+            state_list='\n'.join(generate_str_states_table())))
+
+if __name__ == '__main__':
+    build_tree.chdir_to_root()
+    OUTPUT_FILE_DIR = sys.argv[1] if len(sys.argv) == 2 else "library"
```

Comment:
Why take the output directory as argument and not the output file name?

Reply:
moved to #4981